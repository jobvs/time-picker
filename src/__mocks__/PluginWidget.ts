export function createEditableValue<T>(
    value: T,
    readOnly: boolean,
    validation: string[]
): PluginWidget.EditableValue<T> {
    return {
        value,
        displayValue: "display",
        status: PluginWidget.ValueStatus.Available,
        validation,
        readOnly,
        formatter: null,
        universe: [],
        setValue: () => null,
        setTextValue: () => null,
        setFormatting: () => null,
        setValidation: () => null
    };
}

export function createDynamicValue<T>(value: T): PluginWidget.DynamicValue<T> {
    return {
        value,
        status: PluginWidget.ValueStatus.Available
    };
}

export function createActionValue(execute: () => void): PluginWidget.ActionValue {
    return {
        canExecute: true,
        isExecuting: false,
        execute
    };
}
