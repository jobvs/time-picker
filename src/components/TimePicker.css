.widget-timepicker {
    display: flex;
    outline: none;
}

.widget-timepicker-input {
    flex: 1 0 auto;
}

/* Input height is different on mobile devices */
@media only screen and (-webkit-min-device-pixel-ratio: 0) and (max-device-width: 1024px) {
    .widget-timepicker-input input {
        min-height: 38px;
        line-height: 20px;
    }
}

.widget-timepicker-button {
    flex: 0 1 auto;
    margin-left: 5px;
}

.widget-timepicker .rdtPicker {
    width: auto;
    font-size: 12px;
    border: 1px solid #d1d6e4;
    border-radius: 4px;
    box-shadow: 0 2px 10px 0 rgba(0, 0, 0, 0.06);
    padding-left: 16px;
    padding-right: 16px;
    margin-top: 0;
}

.widget-timepicker .rdtBtn {
    color: #0595db;
    font-size: 11px;
}
