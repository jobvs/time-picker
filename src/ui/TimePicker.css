.widget-timepicker {
  display: flex;
}

.widget-timepicker-input {
  flex: 1 0 auto;
}

.widget-timepicker-button {
  flex: 0 1 auto;
  margin-left: 5px;
}