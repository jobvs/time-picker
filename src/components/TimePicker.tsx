import moment from "moment";
import React from "react";
import Datetime from "react-datetime";
import { hot } from "react-hot-loader";

import "react-datetime/css/react-datetime.css";

import "./TimePicker.css";

interface ActionProps {
    onChange?: PluginWidget.ActionValue;
    onEnter?: PluginWidget.ActionValue;
    onLeave?: PluginWidget.ActionValue;
}

export interface Props extends ActionProps {
    id: string;
    tabIndex?: number;
    timeFormat: "minutes" | "seconds";
    inputValue?: PluginWidget.EditableValue<Date | undefined>;
    editable: "default" | "never";
    placeholder: PluginWidget.DynamicValue<string>;
    validationMessage: PluginWidget.DynamicValue<string>;
}

interface State {
    readonly invalidInput: boolean;
    readonly open: boolean;
}

export class TimePicker extends React.Component<Props, State> {
    readonly state: State = {
        invalidInput: false,
        open: false
    };

    private readonly use12HourNotation = use12HourNotation();
    private readonly useNativeInput = useNativeInput(this.props);

    private blurTimeoutRef: number | undefined;

    constructor(props: Props) {
        super(props);
        this.handleChange = this.handleChange.bind(this);
        this.handleFocus = this.handleFocus.bind(this);
        this.queueBlur = this.queueBlur.bind(this);
        this.handleBlur = this.handleBlur.bind(this);
        this.handleButtonClick = this.handleButtonClick.bind(this);
        this.handleKeyDown = this.handleKeyDown.bind(this);
    }

    get value() {
        return this.props.inputValue && this.props.inputValue.value;
    }

    get validation() {
        return (this.props.inputValue && this.props.inputValue.validation) || [];
    }

    get disabled() {
        return this.props.editable === "never" || !this.props.inputValue || this.props.inputValue.readOnly;
    }

    get placeholder() {
        return !this.props.inputValue
            ? "No attribute selected"
            : this.props.placeholder.value || humanReadableFormat(this.format);
    }

    get format() {
        return timeFormat(this.props.timeFormat, this.use12HourNotation);
    }

    handleFocus() {
        const alreadyFocused = this.blurTimeoutRef !== undefined;

        if (alreadyFocused) {
            clearTimeout(this.blurTimeoutRef);
            this.blurTimeoutRef = undefined;
        } else {
            this.dispatchAction("onEnter");
        }
    }

    queueBlur() {
        if (this.blurTimeoutRef === undefined) {
            this.blurTimeoutRef = setTimeout(() => {
                this.blurTimeoutRef = undefined;
                this.handleBlur();
            }, 0);
        }
    }

    handleBlur() {
        if (this.state.invalidInput && this.props.inputValue) {
            const format = humanReadableFormat(this.format);
            this.props.inputValue.setValidation(
                `${this.props.validationMessage.value || "Expected format:"} ${format}`
            );
            this.setState({ invalidInput: true });
        }

        if (this.state.open) {
            this.setState({ open: false });
        }

        this.dispatchAction("onLeave");
    }

    handleChange(input: string | moment.Moment) {
        if (typeof input === "string") {
            this.handleStringInput(input);
        } else {
            this.setValue(input.toDate());
        }
    }

    handleStringInput(value: string) {
        const invalidInput = value.length > 0;

        if (invalidInput) {
            this.setState({ invalidInput: true });
        } else {
            this.setValue(undefined);
        }
    }

    private setValue(value: Date | undefined) {
        if (!this.props.inputValue) {
            return;
        }

        this.props.inputValue.setValue(value);
        this.props.inputValue.setValidation();
        this.setState({ invalidInput: false });
        this.dispatchAction("onChange");
    }

    handleButtonClick() {
        this.setState({ open: !this.state.open });
    }

    handleKeyDown(event: React.KeyboardEvent) {
        if (event.key === "Escape") {
            this.setState({ open: false });
        }
    }

    render() {
        return (
            <>
                <div
                    className="widget-timepicker"
                    tabIndex={-1}
                    onFocusCapture={this.handleFocus}
                    onBlurCapture={this.queueBlur}
                    onKeyDownCapture={this.handleKeyDown}
                >
                    <Datetime
                        className="widget-timepicker-input"
                        value={this.value}
                        onChange={this.handleChange}
                        dateFormat={false}
                        timeFormat={this.useNativeInput ? "HH:mm" : this.format}
                        open={this.useNativeInput ? false : this.state.open}
                        inputProps={{
                            id: this.props.id,
                            className: "form-control",
                            tabIndex: this.props.tabIndex,
                            placeholder: this.placeholder,
                            disabled: this.disabled,
                            type: this.useNativeInput ? "time" : "text"
                        }}
                    />
                    {this.renderButton()}
                </div>
                {this.renderValidationMessages()}
            </>
        );
    }

    renderButton() {
        return (
            !this.useNativeInput && (
                <button
                    type="button"
                    className="btn mx-button widget-timepicker-button"
                    aria-label="Show time picker"
                    tabIndex={-1}
                    disabled={this.disabled}
                    onClick={this.handleButtonClick}
                >
                    <span className="glyphicon glyphicon-time" />
                </button>
            )
        );
    }

    renderValidationMessages() {
        return this.validation.map(message => (
            <div key={message} className="alert alert-danger mx-validation-message">
                {message}
            </div>
        ));
    }

    private dispatchAction(property: keyof ActionProps): void {
        const prop = this.props[property];
        if (prop && this.props.inputValue && this.props.inputValue.status === PluginWidget.ValueStatus.Available) {
            prop.execute();
        }
    }
}

function useNativeInput(props: Props) {
    return props.timeFormat === "minutes" && isMobileDevice() && supportsNativeInputType("time");
}

function isMobileDevice() {
    return /(iPhone|iPod|iPad|Android|Windows Phone)/.test(navigator.userAgent);
}

function supportsNativeInputType(type: string) {
    const input = document.createElement("input");
    input.type = type;
    input.value = ":)";
    return input.type === type && input.value !== ":)";
}

function timeFormat(detail: "minutes" | "seconds", use12HourNotation: boolean) {
    const hourFormat = use12HourNotation ? "hh" : "HH";
    const seconds = detail === "seconds" ? ":ss" : "";
    const suffix = use12HourNotation ? " A" : "";

    return `${hourFormat}:mm${seconds}${suffix}`;
}

function humanReadableFormat(format: string) {
    return format.replace(" A", " AM/PM");
}

function use12HourNotation() {
    const dojo = (window as any).dojo as { locale: string } | undefined;

    const localesWith12HourNotation = [
        "en-au",
        "en-ca",
        "en-in",
        "en-mt",
        "en-nz",
        "en-ph",
        "en-sg",
        "en-za",
        "en-us",
        "el-cy",
        "el-gr",
        "hi-in",
        "ko-kr",
        "es-do",
        "es-us"
    ];

    return dojo ? localesWith12HourNotation.indexOf(dojo.locale) !== -1 : false;
}

// tslint:disable-next-line:no-default-export
export default hot(module)(TimePicker);
