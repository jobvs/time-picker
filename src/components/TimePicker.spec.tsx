import { mount } from "enzyme";
import React from "react";
import renderer from "react-test-renderer";

import { createActionValue, createDynamicValue, createEditableValue } from "../../tests/mocks/PluginWidget";
import { Props as TimePickerProps, TimePicker } from "./TimePicker";

const setup = (propOverrides: Partial<TimePickerProps> = {}) => {
    const props: TimePickerProps = {
        id: "unique-id",
        inputValue: createEditableValue(new Date(2018, 0, 1, 1, 2, 3, 456), false, []),
        editable: "default",
        timeFormat: "seconds",
        placeholder: createDynamicValue("Placeholder"),
        validationMessage: createDynamicValue("Expected format:"),
        tabIndex: 0,
        onChange: createActionValue(jest.fn),
        onEnter: createActionValue(jest.fn),
        onLeave: createActionValue(jest.fn),
        ...propOverrides
    };

    return <TimePicker {...props} />;
};

describe("TimePicker", () => {
    describe("snapshots", () => {
        it("should render the structure", () => {
            const el = setup();
            const tree = renderer.create(el).toJSON();
            expect(tree).toMatchSnapshot();
        });

        it("should render without a data attribute", () => {
            const el = setup({
                inputValue: undefined
            });
            const tree = renderer.create(el).toJSON();
            expect(tree).toMatchSnapshot();
        });

        it("should render 12-hour notation", () => {
            (window as any).dojo = {
                locale: "en-us"
            };
            const el = setup();
            const tree = renderer.create(el).toJSON();
            expect(tree).toMatchSnapshot();
            (window as any).dojo = undefined;
        });

        it("should render a default human readable placeholder", () => {
            const el = setup({
                placeholder: createDynamicValue("")
            });
            const tree = renderer.create(el).toJSON();
            expect(tree).toMatchSnapshot();
        });

        it("should render validation messages", () => {
            const el = setup({
                inputValue: createEditableValue(new Date(2018, 0, 1, 1, 2, 3, 456), false, ["Error1", "Error2"])
            });
            const tree = renderer.create(el).toJSON();
            expect(tree).toMatchSnapshot();
        });
    });

    describe("onEnter", () => {
        it("is called", () => {
            const mock = jest.fn();
            const el = setup({
                onEnter: createActionValue(mock)
            });
            const wrapper = mount(el);
            wrapper.find("input").simulate("focus");
            expect(mock).toHaveBeenCalledTimes(1);
        });
    });

    describe("onChange", () => {
        it("is called", () => {
            const mock = jest.fn();
            const el = setup({
                onChange: createActionValue(mock)
            });
            const wrapper = mount(el);
            wrapper.find("input").simulate("change");
            expect(mock).toHaveBeenCalledTimes(1);
        });

        it("is not called without inputValue", () => {
            const mock = jest.fn();
            const el = setup({
                onChange: createActionValue(mock),
                inputValue: undefined
            });
            const wrapper = mount(el);
            wrapper.find("input").simulate("change");
            expect(mock).toHaveBeenCalledTimes(0);
        });

        it("is not called with invalid input", () => {
            const mock = jest.fn();
            const el = setup({
                onChange: createActionValue(mock)
            });
            const wrapper = mount(el);
            wrapper.find("input").simulate("change", { target: { value: "asd" } });
            expect(mock).not.toHaveBeenCalled();
        });

        it("is called with empty input", () => {
            const mock = jest.fn();
            const el = setup({
                onChange: createActionValue(mock)
            });
            const wrapper = mount(el);
            wrapper.find("input").simulate("change", { target: { value: "" } });
            expect(mock).toHaveBeenCalledTimes(1);
        });
    });

    // describe("onLeave", () => {
    //     it("is called", () => {
    //         const mock = jest.fn();
    //         const el = setup({ onLeave: createActionValue(mock) });
    //         const wrapper = mount(el);
    //         wrapper.find("input").simulate("focus");
    //         expect(wrapper.find(".rdtPicker").length).toBe(1);
    //         wrapper.simulate("click");
    //         expect(wrapper.find(".rdtPicker").length).toBe(0);
    //         expect(mock).toHaveBeenCalledTimes(1);
    //     });
    // });
});
