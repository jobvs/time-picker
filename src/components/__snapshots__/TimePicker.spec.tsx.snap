// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePicker snapshots should render 12-hour notation 1`] = `
<div
  className="widget-timepicker"
  onBlurCapture={[Function]}
  onFocusCapture={[Function]}
  onKeyDownCapture={[Function]}
  tabIndex={-1}
>
  <div
    className="rdt widget-timepicker-input"
  >
    <input
      className="form-control"
      disabled={false}
      id="unique-id"
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="Placeholder"
      tabIndex={0}
      type="text"
      value="01:02:03 AM"
    />
    <div
      className="rdtPicker"
    >
      <div
        className="rdtTime"
      >
        <table>
          <tbody>
            <tr>
              <td>
                <div
                  className="rdtCounters"
                >
                  <div
                    className="rdtCounter"
                  >
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▲
                    </span>
                    <div
                      className="rdtCount"
                    >
                      1
                    </div>
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▼
                    </span>
                  </div>
                  <div
                    className="rdtCounterSeparator"
                  >
                    :
                  </div>
                  <div
                    className="rdtCounter"
                  >
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▲
                    </span>
                    <div
                      className="rdtCount"
                    >
                      02
                    </div>
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▼
                    </span>
                  </div>
                  <div
                    className="rdtCounterSeparator"
                  >
                    :
                  </div>
                  <div
                    className="rdtCounter"
                  >
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▲
                    </span>
                    <div
                      className="rdtCount"
                    >
                      03
                    </div>
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▼
                    </span>
                  </div>
                  <div
                    className="rdtCounter"
                  >
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▲
                    </span>
                    <div
                      className="rdtCount"
                    >
                      AM
                    </div>
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▼
                    </span>
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <button
    aria-label="Show time picker"
    className="btn mx-button widget-timepicker-button"
    disabled={false}
    onClick={[Function]}
    tabIndex={-1}
    type="button"
  >
    <span
      className="glyphicon glyphicon-time"
    />
  </button>
</div>
`;

exports[`TimePicker snapshots should render a default human readable placeholder 1`] = `
<div
  className="widget-timepicker"
  onBlurCapture={[Function]}
  onFocusCapture={[Function]}
  onKeyDownCapture={[Function]}
  tabIndex={-1}
>
  <div
    className="rdt widget-timepicker-input"
  >
    <input
      className="form-control"
      disabled={false}
      id="unique-id"
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="HH:mm:ss"
      tabIndex={0}
      type="text"
      value="01:02:03"
    />
    <div
      className="rdtPicker"
    >
      <div
        className="rdtTime"
      >
        <table>
          <tbody>
            <tr>
              <td>
                <div
                  className="rdtCounters"
                >
                  <div
                    className="rdtCounter"
                  >
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▲
                    </span>
                    <div
                      className="rdtCount"
                    >
                      1
                    </div>
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▼
                    </span>
                  </div>
                  <div
                    className="rdtCounterSeparator"
                  >
                    :
                  </div>
                  <div
                    className="rdtCounter"
                  >
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▲
                    </span>
                    <div
                      className="rdtCount"
                    >
                      02
                    </div>
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▼
                    </span>
                  </div>
                  <div
                    className="rdtCounterSeparator"
                  >
                    :
                  </div>
                  <div
                    className="rdtCounter"
                  >
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▲
                    </span>
                    <div
                      className="rdtCount"
                    >
                      03
                    </div>
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▼
                    </span>
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <button
    aria-label="Show time picker"
    className="btn mx-button widget-timepicker-button"
    disabled={false}
    onClick={[Function]}
    tabIndex={-1}
    type="button"
  >
    <span
      className="glyphicon glyphicon-time"
    />
  </button>
</div>
`;

exports[`TimePicker snapshots should render the structure 1`] = `
<div
  className="widget-timepicker"
  onBlurCapture={[Function]}
  onFocusCapture={[Function]}
  onKeyDownCapture={[Function]}
  tabIndex={-1}
>
  <div
    className="rdt widget-timepicker-input"
  >
    <input
      className="form-control"
      disabled={false}
      id="unique-id"
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="Placeholder"
      tabIndex={0}
      type="text"
      value="01:02:03"
    />
    <div
      className="rdtPicker"
    >
      <div
        className="rdtTime"
      >
        <table>
          <tbody>
            <tr>
              <td>
                <div
                  className="rdtCounters"
                >
                  <div
                    className="rdtCounter"
                  >
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▲
                    </span>
                    <div
                      className="rdtCount"
                    >
                      1
                    </div>
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▼
                    </span>
                  </div>
                  <div
                    className="rdtCounterSeparator"
                  >
                    :
                  </div>
                  <div
                    className="rdtCounter"
                  >
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▲
                    </span>
                    <div
                      className="rdtCount"
                    >
                      02
                    </div>
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▼
                    </span>
                  </div>
                  <div
                    className="rdtCounterSeparator"
                  >
                    :
                  </div>
                  <div
                    className="rdtCounter"
                  >
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▲
                    </span>
                    <div
                      className="rdtCount"
                    >
                      03
                    </div>
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▼
                    </span>
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <button
    aria-label="Show time picker"
    className="btn mx-button widget-timepicker-button"
    disabled={false}
    onClick={[Function]}
    tabIndex={-1}
    type="button"
  >
    <span
      className="glyphicon glyphicon-time"
    />
  </button>
</div>
`;

exports[`TimePicker snapshots should render validation messages 1`] = `
Array [
  <div
    className="widget-timepicker"
    onBlurCapture={[Function]}
    onFocusCapture={[Function]}
    onKeyDownCapture={[Function]}
    tabIndex={-1}
  >
    <div
      className="rdt widget-timepicker-input"
    >
      <input
        className="form-control"
        disabled={false}
        id="unique-id"
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        placeholder="Placeholder"
        tabIndex={0}
        type="text"
        value="01:02:03"
      />
      <div
        className="rdtPicker"
      >
        <div
          className="rdtTime"
        >
          <table>
            <tbody>
              <tr>
                <td>
                  <div
                    className="rdtCounters"
                  >
                    <div
                      className="rdtCounter"
                    >
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▲
                      </span>
                      <div
                        className="rdtCount"
                      >
                        1
                      </div>
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▼
                      </span>
                    </div>
                    <div
                      className="rdtCounterSeparator"
                    >
                      :
                    </div>
                    <div
                      className="rdtCounter"
                    >
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▲
                      </span>
                      <div
                        className="rdtCount"
                      >
                        02
                      </div>
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▼
                      </span>
                    </div>
                    <div
                      className="rdtCounterSeparator"
                    >
                      :
                    </div>
                    <div
                      className="rdtCounter"
                    >
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▲
                      </span>
                      <div
                        className="rdtCount"
                      >
                        03
                      </div>
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▼
                      </span>
                    </div>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
    <button
      aria-label="Show time picker"
      className="btn mx-button widget-timepicker-button"
      disabled={false}
      onClick={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="glyphicon glyphicon-time"
      />
    </button>
  </div>,
  <div
    className="alert alert-danger mx-validation-message"
  >
    Error1
  </div>,
  <div
    className="alert alert-danger mx-validation-message"
  >
    Error2
  </div>,
]
`;

exports[`TimePicker snapshots should render without a data attribute 1`] = `
<div
  className="widget-timepicker"
  onBlurCapture={[Function]}
  onFocusCapture={[Function]}
  onKeyDownCapture={[Function]}
  tabIndex={-1}
>
  <div
    className="rdt widget-timepicker-input"
  >
    <input
      className="form-control"
      disabled={true}
      id="unique-id"
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="No attribute selected"
      tabIndex={0}
      type="text"
      value=""
    />
    <div
      className="rdtPicker"
    >
      <div
        className="rdtTime"
      >
        <table>
          <tbody>
            <tr>
              <td>
                <div
                  className="rdtCounters"
                >
                  <div
                    className="rdtCounter"
                  >
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▲
                    </span>
                    <div
                      className="rdtCount"
                    >
                      0
                    </div>
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▼
                    </span>
                  </div>
                  <div
                    className="rdtCounterSeparator"
                  >
                    :
                  </div>
                  <div
                    className="rdtCounter"
                  >
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▲
                    </span>
                    <div
                      className="rdtCount"
                    >
                      00
                    </div>
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▼
                    </span>
                  </div>
                  <div
                    className="rdtCounterSeparator"
                  >
                    :
                  </div>
                  <div
                    className="rdtCounter"
                  >
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▲
                    </span>
                    <div
                      className="rdtCount"
                    >
                      00
                    </div>
                    <span
                      className="rdtBtn"
                      onContextMenu={[Function]}
                      onMouseDown={[Function]}
                    >
                      ▼
                    </span>
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <button
    aria-label="Show time picker"
    className="btn mx-button widget-timepicker-button"
    disabled={true}
    onClick={[Function]}
    tabIndex={-1}
    type="button"
  >
    <span
      className="glyphicon glyphicon-time"
    />
  </button>
</div>
`;
